overlord/firstboot.go

```diff
@@ -21,51 +21,80 @@ package overlord
 
 import (
 	"fmt"
+	"os"
+	"path/filepath"
 
 	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/firstboot"
+	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/osutil"
 	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
-	"github.com/snapcore/snapd/snappy"
+	"github.com/snapcore/snapd/snap"
 )
 
 func populateStateFromInstalled() error {
-	all, err := (&snappy.Overlord{}).Installed()
+	if osutil.FileExists(dirs.SnapStateFile) {
+		return fmt.Errorf("cannot create state: state %q already exists", dirs.SnapStateFile)
+	}
+
+	ovld, err := New()
 	if err != nil {
 		return err
 	}
+	ovld.Loop()
+	st := ovld.State()
 
-	if osutil.FileExists(dirs.SnapStateFile) {
-		return fmt.Errorf("cannot create state: state %q already exists", dirs.SnapStateFile)
+	all, err := filepath.Glob(filepath.Join(dirs.SnapBlobDir, "*.snap"))
+	if err != nil {
+		return err
 	}
 
-	st := state.New(&overlordStateBackend{
-		path: dirs.SnapStateFile,
-	})
-	st.Lock()
-	defer st.Unlock()
+	for _, snapPath := range all {
+		sf, err := snap.Open(snapPath)
+		if err != nil {
+			return err
+		}
+		info, err := snap.ReadInfoFromSnapFile(sf, nil)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Installing %s\n", info.Name())
+
+		st.Lock()
+		ts, err := snapstate.InstallPathWithSideInfo(st, info.Name(), snapPath, "", 0)
```

Comment:
also this merit at least a note, if we crash and get back here, because the changes may be in the state we will get conflicts about them